src/builder/builder.go

```diff
@@ -254,33 +254,53 @@ func (b *Builder) AddBundles(bundles []string, force bool, git bool) int {
 		b.UpdateRepo(b.Clearver, false)
 	}
 
-	var includes []string
-	for _, bundle := range bundles {
-		// Check if bundle exists in clrbundledir
-		if _, err := os.Stat(clrbundledir + bundle); os.IsNotExist(err) {
-			helpers.PrintError(errors.New("Bundle " + bundle + " does not exist in CLR version " + b.Clearver))
+	// Copy all bundles over into mix-bundles if -all passed
+	if allbundles {
+		files, err := ioutil.ReadDir(clrbundledir)
+		if err != nil {
+			helpers.PrintError(err)
 			os.Exit(1)
 		}
-		// Check if bundle exists in mix bundles dir
-		if _, err := os.Stat(bundledir + bundle); os.IsNotExist(err) || force {
-			// Parse bundle to get all includes
-			if ib, err := helpers.GetIncludedBundles(clrbundledir + bundle); err != nil {
-				helpers.PrintError(errors.New("Cannot parse bundle " + bundle + " from CLR version " + b.Clearver))
+		bundles = make([]string, len(files))
+		for _, file := range files {
```

Comment:
Instead of duplicating the work, you could append everything to the bundles list here and check if the allBundles flag is set before checking for includes. This keeps the processing of bundles in a single block and remove the need to nest the entire code for the default case. It will make the code easier to maintain.

Reply:
I thought about that, and @mdhorn had the same question.  I did it this way to avoid muddying the already somewhat complicated logic of recursion and file inspection below.  I would still need an `if allbundles` outside of the main loop that copies over all of the file.Name() elements into the bundles slice, and I would need control flow logic to prevent it from parsing the bundles and recursing.

I honestly thought this way would be easier to understand in the future, but since you both commented on it, I'll reconsider the flow.

Reply:
Indeed, makes sense. But in practice it took me much longer to understand the code since the approach I mentioned was my expectation. I had a "wait a minute, whats going on here" moment when reading the code. It is just my impression though. If you believe your approach is better I'm also ok with it.

Reply:
I've given it a go, and the alternative was less intrusive than I had originally anticipated -- I basically get the recursion avoidance for free, because it's already wrapped in a `if len(includes) > 0`.

It was originally my gut instinct to write it that way, and you and @mdhorn both expected it to be that way, so I definitely think it's the right move.

I'll update the PR in a bit when I'm done testing my change.